ocfweb/docs/views/servers.py

```diff
@@ -57,81 +57,89 @@ def has_munin(self):
-                    'democracy',
-                    'firestorm',
-                    'flood',
-                    'lightning',
-                    'maelstrom',
-                    'pestilence',
-                    'reaper',
-                    'supernova',
-                    'thunder',
-                    'tsunami',
-                    'werewolves',
-                    'whirlwind',
-                    'whiteout',
-                    'zombies',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='pandemic',
-            type='hypervisor',
-            children=[
-                Host.from_ldap(hostname)
-                for hostname in (
-                    'pileup',
-                )
-            ],
-        ),
-
-        Host.from_ldap(
-            hostname='riptide',
-            type='hypervisor',
-        ),
-
-        Host.from_ldap(
-            hostname='corruption',
-            type='server',
-        ),
-
-        Host('blackhole', 'network', 'Managed Cisco Catalyst 2960S-48TS-L Switch.', []),
-
-        Host('logjam', 'printer', '', []),
-        Host('pagefault', 'printer', '', []),
-        Host('papercut', 'printer', '', []),
-        Host.from_ldap('overheat', type='raspi'),
-        Host.from_ldap('tornado', type='nuc'),
-
-        *(Host.from_ldap(desktop, type='desktop')
-          for desktop in sorted(list_desktops())),
-    ]
+    servers = create_hosts(hosts_by_filter('(type=server)'))
+    desktops = create_hosts(hosts_by_filter('(type=desktop)'))
+    misc = create_hosts(hosts_by_filter('(type=printer)'))
+    hypervisors = {}
+
+    hypervisor_hostnames = query_puppet(PQL_IS_HYPERVISOR)
+    all_children = query_puppet(PQL_GET_CHILDREN)
+
+    # Add children to hypervisors
+    for h in list(servers.values()):
+        if h.hostname in hypervisor_hostnames:
+            # Populate a list of children
+            children = []
+            for child_hostname in all_children.get(h.hostname, []):
+                child = servers.get(child_hostname)
+                if child:
+                    del servers[child.hostname]
+                    children.append(Host(child.hostname, 'vm', child.description, ()))
+            # Associate host with its children and move it to the hypervisors dictionary
+            del servers[h.hostname]
+            hypervisors[h.hostname] = Host(
+                h.hostname,
+                'hypervisor',
+                h.description,
+                children,
+            )
+
+    # Handle special cases
+    def change_host_type(hostname, host_type, origin_type_dict, target_type_dict):
+        """Pop Host with hostname from servers, change its type, and add to misc
+        """
+        host = origin_type_dict.pop(hostname)
+        target_type_dict[hostname] = host._replace(type=host_type)
+
+    change_host_type('overheat', 'raspi', servers, misc)
+    change_host_type('tornado', 'nuc', servers, misc)
```

Comment:
Nested functions are messy, and this one has a lot of arguments which makes it even more messy. I think the best way to handle this would be at the beginning, when you first interpret the `hosts_by_filter` output.